src/hotspot/share/utilities/resourceHash.hpp

```diff
@@ -257,24 +279,32 @@ class ResourceHashtableBase : public STORAGE {
   // the entry is deleted.
   template<class ITER>
   void unlink(ITER* iter) {
-    const unsigned sz = table_size();
-    for (unsigned index = 0; index < sz; index++) {
-      Node** ptr = bucket_at(index);
-      while (*ptr != nullptr) {
-        Node* node = *ptr;
-        // do_entry must clean up the key and value in Node.
-        bool clean = iter->do_entry(node->_key, node->_value);
-        if (clean) {
-          *ptr = node->_next;
-          if (ALLOC_TYPE == AnyObj::C_HEAP) {
-            delete node;
-          }
-          _number_of_entries --;
-        } else {
-          ptr = &(node->_next);
-        }
+    auto wrapper = [&](Node* const node, Node**& ptr) {
```

Comment:
I wanted unlink to be lambda enabled, not an unlink_impl.  Can you just make unlink_all be something that provides a "return true" lambda and call unlink with that?

Reply:
Or rather, first have a patch to add  unlink that takes a lambda and convert the uses to that.  I don't see the motivation for this change.

Reply:
hi, @coleenp 

My primary goal is 'unlink_all()'. I can create a Simple Iterator and use unlink() to implement that, but it would be sub-optimal(eg. [jvmtiTagMapTable::clear](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/prims/jvmtiTagMapTable.cpp#L71)). When I roll it, I just realize that there are common logic among unlink(), unlink_all() and dtor. 'unlink_impl' is a template. It's not an API here. 

We can provide a lambda version of 'unlink()' in a new JBS issue, but I don't think we need to replace it with unlink(Iter*) completely. Functor and Lambda have different advantages. Functor is more formal and can be stateful. eg. 
`PurgeUnloadedConstraints` is big and deserve indents and comments. 'ResolutionErrorDeleteIterate' captures 'ConstantPool* p' for do_entry(). Can we do this is a separate JBS?

Reply:
My request is to make the unlink APIs use the same design pattern as the iterate APIs. You should be able to use either lambda or a function object for doing the unlink. There should be no performance penalty either way because the C++ compiler can optimize all the noise out.